File: python3/zigzag_conversion.py
"""
Space   : O(n)
Time    : O(n)
"""


class Solution:
    def convert(self, s: str, numRows: int) -> str:
        makeWord = {}

        if numRows == 1:
            return s

        n = len(s)
        i = 0
        count = 1
        direction = '+'

        while i < n:
            if count not in makeWord:
                makeWord[count] = s[i]
            else:
                makeWord[count] += s[i]

            if count == 1:
                direction = '+'
            elif count == numRows:
                direction = '-'

            if direction == '+':
                count += 1
            elif direction == '-':
                count -= 1

            i += 1

        ans = ''
        for i in range(1, numRows+1):
            if i in makeWord:
                ans += makeWord[i]
            else:
                break

        return ans

File: golang/zigzag_conversion.go
package main

// Space	: O(n)
// Time		: O(n)

func convert(s string, numRows int) string {
	ans := ""
	d := map[int]string{}

	if numRows == 1 {
		return s
	}

	count := 1
	direction := "+"

	for _, c := range s {
		if _, ok := d[count]; !ok {
			d[count] = string(c)
		} else {
			d[count] += string(c)
		}

		if direction == "+" && count == numRows {
			direction = "-"
		}
		if direction == "-" && count == 1 {
			direction = "+"
		}

		if direction == "+" {
			count++
		} else if direction == "-" {
			count--
		}
	}

	for i := 1; i <= numRows; i++ {
		ans += d[i]
	}

	return ans
}
